panic in `pulumi convert` for a go program

Looks like there was a panic on the other side of an RPC boundary trying to convert the following YAML into go: https://github.com/pulumi/pulumi-auto-deploy/blob/main/examples/yaml/Pulumi.yaml

Conversion of the same program succeeded for node, python, and csharp.

```
$ pulumi convert --generate-only --out ../go --language go    
Converting from yaml...
Converting to go...
panic: runtime error: invalid memory address or nil pointer dereference

[signal SIGSEGV: segmentation violation code=0x2 addr=0x0 pc=0x105584734]

goroutine 37 [running]:

github.com/pulumi/pulumi/pkg/v3/codegen/go.GenerateProject({0x140000b0148, 0x5}, {{0x140002d9400, 0x10}, {{0x1055a31f4, 0x2}, 0x0}, {0x0, 0x0}, 0x14000409690, ...}, ...)

        /home/runner/work/pulumi/pulumi/pkg/codegen/go/gen_program.go:554 +0x6d4

main.(*goLanguageHost).GenerateProject(0x1400019e3c0, {0x10490bf1c?, 0x40?}, 0x1400060c240)

        /home/runner/work/pulumi/pulumi/sdk/go/pulumi-language-go/main.go:855 +0x360

github.com/pulumi/pulumi/sdk/v3/proto/go._LanguageRuntime_GenerateProject_Handler.func1({0x105fe6f10, 0x140007a74a0}, {0x105e99a00?, 0x1400060c240})

        /home/runner/work/pulumi/pulumi/sdk/proto/go/language_grpc.pb.go:429 +0x74

github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc.OpenTracingServerInterceptor.func1({0x105fe6f10, 0x140007a6420}, {0x105e99a00, 0x1400060c240}, 0x140003182c0, 0x140000b2ba0)

        /home/runner/go/pkg/mod/github.com/grpc-ecosystem/grpc-opentracing@v0.0.0-20180507213350-8e809c8a8645/go/otgrpc/server.go:57 +0x308

github.com/pulumi/pulumi/sdk/v3/proto/go._LanguageRuntime_GenerateProject_Handler({0x105f21e60?, 0x1400019e3c0}, {0x105fe6f10, 0x140007a6420}, 0x14000144150, 0x14000488940)

        /home/runner/work/pulumi/pulumi/sdk/proto/go/language_grpc.pb.go:431 +0x138

google.golang.org/grpc.(*Server).processUnaryRPC(0x1400046c3c0, {0x105fefc00, 0x140006831e0}, 0x140004019e0, 0x14000424ed0, 0x106e48ab0, 0x0)

        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.55.0/server.go:1337 +0xc90

google.golang.org/grpc.(*Server).handleStream(0x1400046c3c0, {0x105fefc00, 0x140006831e0}, 0x140004019e0, 0x0)

        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.55.0/server.go:1714 +0x82c

google.golang.org/grpc.(*Server).serveStreams.func1.1()

        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.55.0/server.go:959 +0x84

created by google.golang.org/grpc.(*Server).serveStreams.func1

        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.55.0/server.go:957 +0x16c

error: could not generate output program: error reading from server: EOF
```

--- a/changelog/pending/20230713--programgen-go--fix-panic-in-generateproject-when-version-is-not-set-in-schema.yaml
+++ b/changelog/pending/20230713--programgen-go--fix-panic-in-generateproject-when-version-is-not-set-in-schema.yaml
@@ -1,0 +1,4 @@
+changes:
+- type: fix
+  scope: programgen/go
+  description: Fix panic in GenerateProject when version is not set in schema

--- a/pkg/codegen/go/gen_program.go
+++ b/pkg/codegen/go/gen_program.go
@@ -455,20 +455,20 @@ func GenerateProgramWithOptions(program *pcl.Program, opts GenerateProgramOption
 	return files, g.diagnostics, nil
 }
 
-func GenerateProject(directory string, project workspace.Project, program *pcl.Program) error {
+func GenerateProjectFiles(project workspace.Project, program *pcl.Program) (map[string][]byte, hcl.Diagnostics, error) {
 	files, diagnostics, err := GenerateProgram(program)
 	if err != nil {
-		return err
+		return files, diagnostics, err
 	}
 	if diagnostics.HasErrors() {
-		return diagnostics
+		return files, diagnostics, err
 	}
 
 	// Set the runtime to "go" then marshal to Pulumi.yaml
 	project.Runtime = workspace.NewProjectRuntimeInfo("go", nil)
 	projectBytes, err := encoding.YAML.Marshal(project)
 	if err != nil {
-		return err
+		return nil, diagnostics, err
 	}
 	files["Pulumi.yaml"] = projectBytes
 
@@ -485,14 +485,14 @@ require (
 	// For each package add a PackageReference line
 	packages, err := programPackageDefs(program)
 	if err != nil {
-		return err
+		return nil, diagnostics, err
 	}
 	for _, p := range packages {
 		if p.Name == "pulumi" {
 			continue
 		}
 		if err := p.ImportLanguages(map[string]schema.Language{"go": Importer}); err != nil {
-			return err
+			return nil, diagnostics, err
 		}
 
 		if p.Version != nil && p.Version.Major <= 0 {
@@ -550,14 +550,31 @@ require (
 			}
 		}
 
+		version := ""
+		if p.Version != nil {
+			version = fmt.Sprintf("v%s", p.Version.String())
+		}
 		if packageName != "" {
-			fmt.Fprintf(&gomod, "	%s v%s\n", packageName, p.Version.String())
+			fmt.Fprintf(&gomod, "	%s %s\n", packageName, version)
 		}
 	}
 
 	gomod.WriteString(")")
 
 	files["go.mod"] = gomod.Bytes()
+
+	return files, diagnostics, nil
+}
+
+func GenerateProject(directory string, project workspace.Project, program *pcl.Program) error {
+	files, diagnostics, err := GenerateProjectFiles(project, program)
+	if err != nil {
+		return err
+	}
+
+	if diagnostics.HasErrors() {
+		return diagnostics
+	}
 
 	for filename, data := range files {
 		outPath := path.Join(directory, filename)

--- a/pkg/codegen/go/gen_program_test.go
+++ b/pkg/codegen/go/gen_program_test.go
@@ -4,7 +4,10 @@ import (
 	"bytes"
 	"os"
 	"path/filepath"
+	"strings"
 	"testing"
+
+	"github.com/pulumi/pulumi/sdk/v3/go/common/workspace"
 
 	"github.com/hashicorp/hcl/v2"
 	"github.com/stretchr/testify/assert"
@@ -127,4 +130,40 @@ func newTestGenerator(t *testing.T, testFile string) *generator {
 	}
 	g.Formatter = format.NewFormatter(g)
 	return g
+}
+
+func parseAndBindProgram(t *testing.T,
+	text string,
+	name string,
+	options ...pcl.BindOption,
+) (*pcl.Program, hcl.Diagnostics, error) {
+	parser := syntax.NewParser()
+	err := parser.ParseFile(strings.NewReader(text), name)
+	if err != nil {
+		t.Fatalf("could not read %v: %v", name, err)
+	}
+	if parser.Diagnostics.HasErrors() {
+		t.Fatalf("failed to parse files: %v", parser.Diagnostics)
+	}
+
+	options = append(options, pcl.PluginHost(utils.NewHost(testdataPath)))
+
+	return pcl.BindProgram(parser.Files, options...)
+}
+
+func TestGenerateProjectDoesNotPanicWhenMissingVersion(t *testing.T) {
+	t.Parallel()
+
+	source := `	
+resource main "auto-deploy:index:AutoDeployer" {
+    project = "example"
+}`
+
+	program, diags, err := parseAndBindProgram(t, source, "main.pp")
+
+	require.NoError(t, err)
+	require.False(t, diags.HasErrors())
+
+	files, diags, err := GenerateProjectFiles(workspace.Project{}, program)
+	assert.NotNil(t, files, "Files were generated")
 }

--- a/pkg/codegen/go/gen_program_test.go
+++ b/pkg/codegen/go/gen_program_test.go
@@ -166,4 +166,6 @@ resource main "auto-deploy:index:AutoDeployer" {
 
 	files, diags, err := GenerateProjectFiles(workspace.Project{}, program)
 	assert.NotNil(t, files, "Files were generated")
+	require.NoError(t, err)
+	require.False(t, diags.HasErrors())
 }

--- a/pkg/codegen/testing/test/testdata/auto-deploy-0.0.1.json
+++ b/pkg/codegen/testing/test/testdata/auto-deploy-0.0.1.json
@@ -1,0 +1,30 @@
+{
+  "name": "auto-deploy",
+  "meta": {
+    "moduleFormat": "(.*)"
+  },
+  "config": {},
+  "provider": {
+    "type": "object"
+  },
+  "resources": {
+    "auto-deploy:index:AutoDeployer": {
+      "description": "Automatically trigger downstream updates on dependent stacks via Pulumi Deployments.\nAutoDeployer requires that stacks have Deployment Settings configured.",
+      "properties": {
+        "project": {
+          "type": "string",
+          "description": "The project name for the AutoDeployer stack."
+        }
+      },
+      "type": "object",
+      "inputProperties": {
+        "project": {
+          "type": "string",
+          "description": "The project name for the AutoDeployer stack."
+        }
+      },
+      "requiredInputs": ["project"],
+      "isComponent": true
+    }
+  }
+}

--- a/pkg/codegen/testing/utils/host.go
+++ b/pkg/codegen/testing/utils/host.go
@@ -86,5 +86,6 @@ func NewHost(schemaDirectoryPath string) plugin.Host {
 		SchemaProvider{"splat", "1.0.0"},
 		SchemaProvider{"snowflake", "0.66.1"},
 		SchemaProvider{"using-dashes", "1.0.0"},
+		SchemaProvider{"auto-deploy", "0.0.1"},
 	)
 }

